luminous-lightyears/src/game.py

```diff
@@ -0,0 +1,210 @@
+        self.game_stop_flag = True
+
+    async def loop(self) -> None:
+        """Define the main loop of the game."""
+        self.start_time = time.time()
+        players = self.players.values()
+        await self.send_stats()
+
+        while True:
+            logger.info(f"{len(self.players)} left in game {self.id}")
+
+            if self.game_stop_flag:
+                break
+
+            game_time: float = (time.time() - self.start_time) / 60
+            if (game_time > self.cumm_percent_time_per_stage[self.stage - 1] * self.max_time) and (
+                game_time < self.max_time
+            ):
+                self.stage = total_stages[total_stages.index(self.stage) + 1]
+                await self.send_stats()
+
+            if game_time >= self.max_time:
+                logger.info(f"Time is Up! Game {self.id} is over!")
+                self.stop()
+                await self.stop_game_by_time()
+                break
+
+            logger.info(f"{game_time=} {self.stage=} {self.max_time=}")
+
+            try:
+                response = await asyncio.gather(*[self.tick(player) for player in players], return_exceptions=True)
+
+                for res in response:
+                    if isinstance(res, Exception):
+                        logger.error(res)
+            except Exception as e:
+                logger.exception("Error occurred in game loop")
+
+                for player in players:
+                    await player.ctx.send(
+                        embed=Embed(
+                            title="Some error occured",
+                            description=f"{e} \n has occured, please contact the devs if you see this",
+                            color=error_color,
+                        ),
+                    )
+
+    async def tick(self, player: Player) -> None:
+        """Define the activities done in every game tick."""
+        if self.game_stop_flag:
+            return
+
+        if (time.time() - player.last_activity_time) > AFK_TIME:
+            await self.disqualify_player(player)
+            return
+
+        character = all_characters.get_random(player.state)
+        for attr in self.values_to_check:
+            if getattr(player.state, attr) <= 0:
+                # Some value is negative hence need to send the losing message
+                await self.death_player(player)
+                return
+
+        # The sleep times are subject to change, based on how the actual gameplay feels
+        # The randomness gives a variability between the values mentioned in the brackets
+        match self.stage:
+            case 1:
+                sleep_time = 10 + (random.uniform(-2, 2))
+            case 2:
+                sleep_time = 8 + (random.uniform(-2, 1.5))
+            case 3:
+                sleep_time = 6 + (random.uniform(-1, 0.75))
+
+        character = all_characters.get_random(player.state)
+        while self.stage not in character.stages:
```

Comment:
As far as I can tell there is no guarantee that this while loop will ever complete. I.e if there is a single character that causes this loop to repeat, it could be called indefinitely until the heat death of the universe. Similar logic is implemented in the `get_random` method. 

Instead, it would be preferable to first build a list of valid choices and select from those valid options, instead of trying forever to find a valid option. 

This code also begs the question "what if there are no valid selections remaining?" Would it be desirable to repeat forever if that were to happen?